plugins/woocommerce/tests/api-core-tests/data/countries/ad.json

```diff
@@ -0,0 +1,17 @@
+{
+    "code": "AD",
+    "name": "Andorra",
+    "states": [],
+    "_links": {
+        "self": [
+            {
+                "href": "http://localhost:8086/wp-json/wc/v3/data/countries/ad"
```

Comment:
Since we've added support for different hosts on https://github.com/woocommerce/woocommerce/pull/35522 we should make sure these have the same flexibility built in. Since these are JSON it might be a bit challenging to try and make this dynamic, so we may want to remove the checks on `_links` for now, as this fails against other environments such as JN:

<img width="1001" alt="Screenshot 2022-11-21 at 10 52 54" src="https://user-images.githubusercontent.com/71906536/203125948-ccf1435a-0727-4b1c-a42b-336a52c37e9b.png">

Reply:
Thanks for taking a look at this @zhongruige - appreciate the catch! 

The tests are actually failing due to the differences in the country data between local and JN in countries such as Senegal and Ukraine.
e.g.
Local Senegal:
```
       "code": "SN",
        "name": "Senegal",
        "states": [
            {
                "code": "SNDB",
                "name": "Diourbel"
            },
            {
                "code": "SNDK",
                "name": "Dakar"
            },
```

JN Senegal:
```
"code": "SN",
        "name": "Senegal",
        "states": [],
```


Local Ukraine:
```
"code": "UA",
        "name": "Ukraine",
        "states": [
            {
                "code": "UA05",
                "name": "Vinnychchyna"
            },
            {
                "code": "UA07",
                "name": "Volyn"
            },
```

JN Ukraine:
```
        "code": "UA",
        "name": "Ukraine",
        "states": [
            {
                "code": "VN",
                "name": "Vinnytsia Oblast"
            },
            {
                "code": "VL",
                "name": "Volyn Oblast"
            },
```

There is obviously a slight disjoin in the data that made it into the release that fuels JN. I can tweak this to make the 'offending' countries more general but it also means that any similar changes in the future will cause similar failures..

The links are not the reason for the test failures here. I had built flexibility into the tests so that the links are not a factor. Here is an example snippet from the test that allows for any baseURL:

```
"_links": {
		"self": [{
			"href": expect.stringContaining(`/wp-json/wc/v3/data/countries/${country}`)
		}],
		"collection": [{
			"href": expect.stringContaining('/wp-json/wc/v3/data/countries')
		}]
	}
```

We could remove the localhost references in the country data files if this would avoid confusion? e.g. replace the `http://localhost:8086` with baseURL or similar - or entirely remove the links sections from the country data files - interested to hear your thoughts on this 😊

Reply:
For now, I will tweak the states test to allow the more generic tests and remove the specific states tests for Senegal and Ukraine.

Reply:
>  allow the more generic tests and remove the specific states tests for Senegal and Ukraine

Ah yeah I can see where these were added and slated for the 7.2.0 release:

* Senegal: https://github.com/woocommerce/woocommerce/pull/35199
* Ukraine: https://github.com/woocommerce/woocommerce/pull/35493

It's possible we may have similar cases in the future and need to update other countries with similar generic checks. I worry that may end up being more of a maintenance overhead for the tests to do regular updates like this. It also leaves a bit of a gap where some locales have more extensive testing for the states, while others are generic. So I'm curious on your thoughts--should we make _all_ the tests generic, or find a way to handle running the tests against where we may have failures depending on the WooCommerce version for added/removed data points? Both definitely present their challenges so I'm interested in what your thoughts are.

Reply:
> This is a good one to ponder indeed. I would personally prefer that we had more tests covered rather than less. The state based tests act as a regression.

Agreed on this point, since testing we get the states field back and it being empty won't tell us if there's an issue where suddenly we're _not_ returning states for a given country due to an error somewhere else.

>  in what situation would we be seeing developers pull the tests from trunk and then run the tests against an environment (such as JN) that was different to the latest that resides in trunk?

Solid point, since I can see a use-case where they might want to run it against the latest released (or pre-release) WooCommerce with, perhaps, an extension they're developing. I don't know if many developers would use the `trunk` version--since that is the least stable and more likely to change--and instead focus in on released versions, so testing against betas and RCs, then the final released version. Perhaps earliest they may try is the nightly version.

> We could theoretically note states that are 'in transition' between release and trunk and avoid executing those state tests until what was released 'caught up' with trunk

One approach I think for this could be to use the [API to get the WC version](https://woocommerce.github.io/woocommerce-rest-api-docs/#system-status-properties) and use that version as a conditional in the tests themselves (something like this could be a useful approach in the future, too, for feature flagging etc). So something like `if ( version > "7.2.0" ) { runTests() }  else { skipTests() }`. This could be ugly to figure out dealing with semver comparisons, but mostly throwing this out as an idea.

Still, this does make me think a bit, I wonder if once we get the API tests into a package, we can set up a regular release matching the WC release to limit the gap. So maybe for now it's actually okay to keep the states in, since this will work in 7.2 eventually, and once we have a releasable package we can adjust? So, the trunk version of the API core tests would work in trunk, but the released version of the API tests package would work on JN sites on the released version of WC.

> Another alternative would be documentation of the process to ensure execution of the appropriate tests i.e. those that related to the version of Woo you were running against.

I do like this, though, either way, documentation is always a win in my book!

Reply:
> I don't know if many developers would use the trunk version--since that is the least stable and more likely to change--and instead focus in on released versions, so testing against betas and RCs, then the final released version. Perhaps earliest they may try is the nightly version.

Interesting, though I assume the nightly version would be very similar to trunk from our perspective..

> One approach I think for this could be to use the [API to get the WC version](https://woocommerce.github.io/woocommerce-rest-api-docs/#system-status-properties) and use that version as a conditional in the tests themselves

I like this consideration 👍 
I wouldn't imagine there would be a truckload of regular updates (correct me if you think otherwise!) so any updates could be made quite quickly if we had settled on a method/pattern - however, as you mention this could be an ugly one to figure out. 😊

> I wonder if once we get the API tests into a package, we can set up a regular release matching the WC release to limit the gap.

Yeah - plus this would mean we are working with a definitive version that won't have to handle possible upcoming changes between release and trunk (and therefore a slight reduction in maintenance). One 'released' version and one operating against the trunk version.

> So maybe for now it's actually okay to keep the states in, since this will work in 7.2 eventually, and once we have a releasable package we can adjust? 

This sounds good to me. I'll make the appropriate updates to ensure all countries that have states are getting tested in trunk - as currently these tests are executing in CI and by the likes of ourselves locally.

> So, the trunk version of the API core tests would work in trunk, but the released version of the API tests package would work on JN sites on the released version of WC.

This seems a good pattern/method to me. Tests can then be updated for trunk as a release progresses and we can handle whatever makes it into a release in the released package version. Thanks for the input!

Reply:
Sounds great! Let's go with this implementation and we'll circle back on the best approach once we get this to a releasable package state. Thank you for the great discussion around this @nigeljamesstevenson!